contracts/strategies/traderjoe/TraderJoe.sol

```diff
@@ -245,41 +219,335 @@ contract TraderJoe is UUPSUpgradeable, StrategyOwnablePausableBaseUpgradeable {
+        ITraderJoeLBPair lbPair,
+        ITraderJoeLBRouter lbRouter,
+        uint256 binStep,
+        uint256[] calldata binIds,
+        uint256[] calldata binAllocations
+    ) private {
+        TraderJoeStorage storage strategyStorage = TraderJoeStorageLib
+            .getStorage();
+
+        strategyStorage.lbPair = lbPair;
+        strategyStorage.lbRouter = lbRouter;
+        strategyStorage.tokenX = IERC20Upgradeable(lbPair.tokenX());
+        strategyStorage.tokenY = IERC20Upgradeable(lbPair.tokenY());
+        strategyStorage.binStep = binStep;
+        strategyStorage.binIds = binIds;
+        strategyStorage.binAllocations = binAllocations;
+
+        if (strategyStorage.tokenX == depositToken) {
+            strategyStorage.pairDepositToken = IERC20Upgradeable(
+                strategyStorage.tokenY
+            );
+        } else if (strategyStorage.tokenY == depositToken) {
+            strategyStorage.pairDepositToken = IERC20Upgradeable(
+                strategyStorage.tokenX
+            );
+        } else {
+            revert InvalidTraderJoeLBPair();
+        }
+    }
+
+    function __deposit(
+        uint256 depositTokenAmount,
+        uint256 pairDepositTokenAmount
+    ) private {
+        TraderJoeStorage storage strategyStorage = TraderJoeStorageLib
+            .getStorage();
+
+        uint256 pairDepositTokenBefore = strategyStorage
+            .pairDepositToken
+            .balanceOf(address(this)) - pairDepositTokenAmount;
+
+        uint256 amountXIn;
+        uint256 amountYIn;
+
+        if (strategyStorage.tokenX == depositToken) {
+            amountXIn = depositTokenAmount;
+            amountYIn = pairDepositTokenAmount;
+        } else {
+            amountXIn = pairDepositTokenAmount;
+            amountYIn = depositTokenAmount;
+        }
+
+        uint256 binsAmount = strategyStorage.binIds.length;
+
+        (, , uint256 activeId) = strategyStorage.lbPair.getReservesAndId();
+
+        uint256 totalAmount = amountXIn + amountYIn;
+        uint256 amountX;
+        uint256 amountY;
+
+        // Distributions.
+        uint256[] memory distributionX = new uint256[](binsAmount);
+        uint256[] memory distributionY = new uint256[](binsAmount);
+
+        for (uint256 i; i < binsAmount; ++i) {
+            // Bin allocation has precision of 1e3.
+            uint256 amount = (totalAmount * strategyStorage.binAllocations[i]) /
+                1e3;
+
+            if (strategyStorage.binIds[i] <= activeId) {
+                distributionY[i] = amount;
+                amountY += amount;
+            } else if (strategyStorage.binIds[i] > activeId) {
+                distributionX[i] = amount;
+                amountX += amount;
+            }
+        }
+
+        for (uint256 i; i < binsAmount; ++i) {
+            // TraderJoe V2 has precision of 1e18. Calibrate distributions so that the sum of them equals to 1e18.
+            if (strategyStorage.binIds[i] <= activeId) {
```

Comment:
question: in what ratio should liquidity be provided to the active bin? Let's say the current price is 1 AssetX worth 2 AssetY. Do I have to provide two times as much AssetY?

Reply:
The ratio of two assets at the moment the tx is being executed. You can swap to match the ratio yourself but you don't have to. If your provided assets are off, TJ imposes swap fee for the difference. (No actual swap occurs but just swap fee is imposed.)

Reply:
so we are double charged on swap fees for deposits on the active bin (for one of the assets), if assetY != USDC(deposit currency), right?

Reply:
cause we swapped too much into assetY

Reply:
can you check if assetY is USDC?
If it is, then we can just make an other comment about the assumption that assetY is USDC
It if is not, we can just ask Jasa if he is okay with this double charging (as we put most of our money into the active bin)

Reply:
Asset Y is USDC and I changed an algorithm of allocating liquidity to an active bin. I'll add tests and update PR tomorrow.